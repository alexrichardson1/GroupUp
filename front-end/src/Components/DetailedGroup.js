import React from "react";
import { useParams } from "react-router-dom";
import "../../node_modules/bootstrap/dist/css/bootstrap.css";
import data from "../Teams";

const DetailedGroup = () => {
  const { id } = useParams();
  const groupId = parseInt(id);
  const { leader, maxMembers, teammates, requirements, adRequirements } =
    data.filter((group) => group.id === groupId)[0];
  const membersNeeded = maxMembers - teammates.length;
  console.log(maxMembers, teammates.length);
  const leaderFirstName = leader.split(" ")[0];

  return (
    <React.Fragment>
      <h1>{groupId}</h1>
      <h1>{leaderFirstName}'s Group</h1>
      <h2>{membersNeeded} members needed</h2>
      <div>
        <h2>Other Members</h2>
        <ul class="list-group">
          {teammates.map((person) => (
            <li class="list-group-item">{person}</li>
          ))}
        </ul>
      </div>
      <h2>Requirements:</h2>
      <h3>Language: {requirements["Code Language"]}</h3>
      <h3>Timezone: GMT{requirements["Timezone"]}</h3>
      <h3>Preferred Language: {requirements["Spoken Language"]}</h3>
      {adRequirements && (
        <div>
          <h3>Additional Requirements:</h3>
          <p>{adRequirements}</p>
        </div>
      )}
      <a href="/listings">
        <button className="btn">Go Back</button>
      </a>
      <button
        type="button"
        class="btn btn-dark"
        onMouseOver={() => alert("Feature not implemented")}
      >
        Request to Join
      </button>
    </React.Fragment>
  );
};

export default DetailedGroup;
