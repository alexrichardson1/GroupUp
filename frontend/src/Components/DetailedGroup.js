import React from "react";
import { useParams } from "react-router-dom";
import "../../node_modules/bootstrap/dist/css/bootstrap.css";
import data from "../Teams";
import { Button, ListGroup } from "react-bootstrap";
import { LinkContainer } from "react-router-bootstrap";
import NavBar from "./NavBar";

const DetailedGroup = () => {
  const { id } = useParams();
  const groupId = parseInt(id);
  const {
    leader,
    maxMembers,
    teammates,
    requirements,
    adRequirements,
    projectId,
  } = data.filter((group) => group.id === groupId)[0];
  const membersNeeded = maxMembers - teammates.length;
  console.log(maxMembers, teammates.length);
  const leaderFirstName = leader.split(" ")[0];
  const goBackLink = "/listing/" + projectId;

  return (
    <React.Fragment>
      <NavBar
        renderBool={[true, true, true, true]}
        id={projectId}
        create={false}
      />
      <h1>{leaderFirstName}'s Group</h1>
      <p>{membersNeeded} members needed.</p>
      <ListGroup variant="flush">
        <ListGroup.Item variant="dark">Requirements:</ListGroup.Item>
        {Object.entries(requirements).map(([key, val]) => (
          <ListGroup.Item>
            {key}: {val}
          </ListGroup.Item>
        ))}
      </ListGroup>
      <ListGroup.Item variant="dark">Other Members:</ListGroup.Item>
      <ListGroup variant="flush">
        {teammates.map((person) => (
          <ListGroup.Item>{person}</ListGroup.Item>
        ))}
      </ListGroup>
      {adRequirements && (
        <div>
          <h5>Additional Requirements:</h5>
          <p>{adRequirements}</p>
        </div>
      )}
      <LinkContainer to={goBackLink}>
        <Button>Go Back</Button>
      </LinkContainer>
      <Button>Show groups</Button>
    </React.Fragment>
  );
};

export default DetailedGroup;
