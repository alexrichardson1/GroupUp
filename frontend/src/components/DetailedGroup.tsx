import React, { useContext } from "react";
import { useParams, useHistory } from "react-router-dom";
import { useState, useEffect } from "react";
import "bootstrap/dist/css/bootstrap.css";
import { Button, ListGroup, Table, Modal } from "react-bootstrap";
import NavBar from "components/NavBar";
import axios from "axios";
import { config } from "Constants";
import { UserContext } from "components/auth/UserContext";
import { ProjectT } from "types/types";
import { dummyProject } from "api";

const DetailedGroup = () => {
  const { id, projectId } = useParams<{ id: string; projectId: string }>();
  const groupId = parseInt(id);
  const history = useHistory();
  const [email, setEmail] = useState<String>("");
  const [leader, setLeader] = useState<String>("");
  const [maxMembers, setMaxMembers] = useState<Number>(0);
  const [teammates, setTeammates] = useState<String[]>([]);
  const [requirements, setRequirements] = useState<string[]>([]); // requirement values: e.g. Java, +1
  const [adRequirements, setAdRequirements] = useState<String>("");
  const [project, setProject] = useState(dummyProject);

  const [show, setShow] = useState(false);
  const { user } = useContext(UserContext);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  console.log(maxMembers, teammates.length);
  const leaderFirstName = leader.split(" ")[0];

  const joinGroup = async () => {
    var result = {};
    await axios
      .post(`${config.API_URL}/group/join`, {
        groupid: groupId,
        name: user,
      })
      .then((res) => {
        const group = res.data;
        result = group;
        alert("Successfully joined group!");
        setTeammates([...teammates, user]);
      })
      .catch((error) => {
        console.log(error);
      });
    return result;
  };

  //@ts-ignore
  useEffect(async () => {
    const getGroup = async () => {
      await axios
        .post(`${config.API_URL}/group/one`, {
          groupid: groupId,
        })
        .then((res) => {
          const group = res.data;
          console.log(group);
          setLeader(group.leader);
          setMaxMembers(group.maxmembers);
          setTeammates(group.teammates);
          setRequirements(group.requirements);
          setAdRequirements(group.adrequirements);
          setEmail(group.leaderemail);
        })
        .catch((error) => {
          console.log(error);
        });
    };

    const getProject = async () => {
      var result: ProjectT = dummyProject;
      await axios
        .get(`${config.API_URL}/project`)
        .then((res) => {
          const projects: ProjectT[] = res.data;
          const filteredProjects = projects.filter(
            (proj) => proj.id === parseInt(projectId)
          );
          if (filteredProjects[0]) {
            result = filteredProjects[0];
          }
          setProject(result);
        })
        .catch((error) => {
          console.error(error);
        });
    };

    document.title = `${leader}'s Group`;
    await getGroup();
    await getProject();
  }, []);

  const isJoined = () => {
    return leader === user || teammates.some((name) => name === user);
  };

  return (
    <React.Fragment>
      <NavBar
        renderBool={[true, true, true, true]}
        id={parseInt(projectId)}
        create={false}
      />
      <Modal show={show} onHide={handleClose}>
        <Modal.Header>
          <Modal.Title>Contact Details</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <p>Leader Email: {email}.</p>
        </Modal.Body>

        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Close
          </Button>
        </Modal.Footer>
      </Modal>
      <h1>{leaderFirstName}'s Group</h1>
      <div>1 members needed.</div>
      <Table striped bordered hover size="sm">
        <thead>
          <tr>
            <th>Requirement</th>
            <th>Needed</th>
          </tr>
        </thead>
        <tbody>
          {Object.entries(requirements).map(([key, val]: [string, string]) => (
            <tr>
              <td>
                {
                  //@ts-ignore
                  project.requirements[key] === "Timezone"
                    ? "Timezone (UTC +/-)"
                    : //@ts-ignore
                      project.requirements[key]
                }
              </td>
              <td>
                {
                  //@ts-ignore
                  project.requirements[key] === "Timezone"
                    ? parseInt(val) > -1
                      ? "+" + val
                      : val
                    : val
                }
              </td>
            </tr>
          ))}
        </tbody>
      </Table>
      <ListGroup.Item variant="dark">Other Members:</ListGroup.Item>
      <ListGroup variant="flush">
        {teammates.map((person) => (
          <ListGroup.Item>{person}</ListGroup.Item>
        ))}
      </ListGroup>
      {adRequirements && (
        <div>
          <h5>Additional Requirements:</h5>
          <div>{adRequirements}</div>
        </div>
      )}
      <Button onClick={() => history.goBack()}>Go Back</Button>
      <Button onClick={handleShow}>Contact Leader</Button>
      <Button
        onClick={() => {
          if (user) joinGroup();
        }}
        disabled={isJoined()}
      >
        {isJoined() ? "Already joined" : "Join Group"}
      </Button>
    </React.Fragment>
  );
};

export default DetailedGroup;
