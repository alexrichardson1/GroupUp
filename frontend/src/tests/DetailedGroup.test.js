import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import DetailedGroup from "components/DetailedGroup";

test("renders correctly", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <DetailedGroup />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
