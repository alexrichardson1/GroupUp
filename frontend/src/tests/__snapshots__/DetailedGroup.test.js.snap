// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <nav
    className="navBar navbar navbar-expand-lg navbar-light fixed-top"
  >
    <span
      className="navbar-brand"
      style={
        Object {
          "color": "white",
        }
      }
    >
      <img
        alt="group up logo"
        className="logoImage"
        href="/home"
        onClick={[Function]}
        src="https://i.imgur.com/w92t1xH.png"
      />
    </span>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
    >
      <div
        className="mr=auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rb-event-key="/home"
          href="/home"
          onClick={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Home
          <img
            alt="drop down arrow"
            className="processImage"
            src="/right-caret.svg"
          />
        </a>
        <a
          className="nav-link"
          data-rb-event-key="/selection"
          href="/selection"
          onClick={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Select Hackathon
          <img
            alt="drop down arrow"
            className="processImage"
            src="/right-caret.svg"
          />
        </a>
        <a
          className="nav-link"
          data-rb-event-key="/listing/NaN"
          href="/listing/NaN"
          onClick={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Group Listings
          <img
            alt="drop down arrow"
            className="processImage"
            src="/right-caret.svg"
          />
        </a>
        <span
          className="navbar-brand"
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Join a group
        </span>
      </div>
      <div
        className="ms-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="ml-auto navHelp navbar-brand"
          href="/help"
          id="collasible-nav-dropdown"
          onClick={[Function]}
          title="Help"
        >
          <img
            alt="help"
            className="splitterImage"
            src="/info-circle.svg"
          />
        </a>
      </div>
    </div>
  </nav>,
  <h1>
    
    's Group
  </h1>,
  <div>
    1 members needed.
  </div>,
  <table
    className="table table-sm table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Requirement
        </th>
        <th>
          Needed
        </th>
      </tr>
    </thead>
    <tbody />
  </table>,
  <div
    className="list-group-item list-group-item-dark"
    disabled={false}
    onClick={[Function]}
  >
    Other Members:
  </div>,
  <div
    className="list-group list-group-flush"
    onKeyDown={[Function]}
  />,
  "",
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Go Back
  </button>,
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Contact Leader
  </button>,
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Join Group
  </button>,
]
`;
